lib/puma/request.rb

```diff
@@ -203,39 +225,102 @@ def default_server_port(env)
       end
     end
 
-    # Writes to an io (normally Client#io) using #syswrite
-    # @param io [#syswrite] the io to write to
+    # Used to write 'early hints', 'no body' responses, 'hijacked' responses,
+    # and body segments (called by `fast_write_response`).
+    # Writes a string to an io (normally `Client#io`) using `write_nonblock`.
+    # Large strings may not be written in one pass, especially if `io` is a
+    # `MiniSSL::Socket`.
+    # @param io [#write_nonblock] the io to write to
     # @param str [String] the string written to the io
     # @raise [ConnectionError]
     #
-    def fast_write(io, str)
+    def fast_write_str(socket, str)
       n = 0
-      while true
+      byte_size = str.bytesize
+      while n < byte_size
         begin
-          n = io.syswrite str
+          n += socket.syswrite(n.zero? ? str : str.byteslice(n..-1))
         rescue Errno::EAGAIN, Errno::EWOULDBLOCK
-          unless io.wait_writable WRITE_TIMEOUT
-            raise ConnectionError, "Socket timeout writing data"
+          unless socket.wait_writable WRITE_TIMEOUT
+            raise ConnectionError, SOCKET_WRITE_ERR_MSG
           end
-
           retry
         rescue  Errno::EPIPE, SystemCallError, IOError
-          raise ConnectionError, "Socket timeout writing data"
+          raise ConnectionError, SOCKET_WRITE_ERR_MSG
         end
-
-        return if n == str.bytesize
-        str = str.byteslice(n..-1)
       end
     end
-    private :fast_write
 
-    # @param status [Integer] status from the app
-    # @return [String] the text description from Puma::HTTP_STATUS_CODES
+    # Used to write headers and body.
+    # Writes to a socket (normally `Client#io`) using `#fast_write_str`.
+    # Accumulates `body` items into `io_buffer`, then writes to socket.
+    # @param socket [#write] the response socket
+    # @param body [Enumerable, File] the body object
+    # @param io_buffer [Puma::IOBuffer] contains headers
+    # @param chunk [Boolean]
+    # @raise [ConnectionError]
     #
-    def fetch_status_code(status)
-      HTTP_STATUS_CODES.fetch(status) { 'CUSTOM' }
+    def fast_write_response(socket, body, io_buffer, chunked)
+      if body.is_a?(::File) || body.respond_to?(:read) || body.respond_to?(:readpartial)
+        if chunked  # would this ever happen?
+          while part = body.read(BODY_LEN_MAX)
+            io_buffer.append part.bytesize.to_s(16), LINE_END, part, LINE_END
+          end
+          io_buffer << CLOSE_CHUNKED
+          fast_write_str socket, io_buffer.to_s
+        else
+          if body.respond_to?(:size) && body.size <= 10_240
+            IO.copy_stream body, io_buffer
+            fast_write_str socket, io_buffer.to_s
+          else
+            fast_write_str socket, io_buffer.to_s
+            IO.copy_stream body, socket
+          end
+        end
+        body.close unless body.closed?
+      elsif body.respond_to?(:to_ary) && body.length == 1
```

Comment:
I'm getting an exception here when the response coming from Rails is already chunked (Rails app sets `Transfer-Encoding: chunked` header), in this case `body` here is a `RackBody` instance that responds to `to_ary` but don't implements `length`:

```
2022-09-05 01:25:12 -0700 Read: #<NoMethodError: undefined method `length' for #<ActionDispatch::Response::RackBody:0x000000014ed57948>>
```

Reply:
Thanks for checking/testing.

Fixed - changed `body.respond_to?(:to_ary) ` to `body.is_a?(::Array)

Maybe the PR needs more tests with `RackBody`.  I'll see about adding some...

Reply:
See additional commit [test_rack_server.rb - add tests for Rack::Chunked middleware](https://github.com/puma/puma/pull/2896/commits/894bf4cd18b795d152c685c7c124392d0cbe31ab?diff=split&w=0)